### DISCUSSION TITLE
Brainvisa singularity using google Colab or Jupiter?

### DISCUSSION THREAD

**Discussion by ece7048** [brainvisa/brainvisa.github.io#119] [Q&A]
Dear all,

I am trying to use the singularity developer source code building in google Colab. As ai need to test some code before I run it on clusters. 
When I am trying to use : **!singularity run -c -B ./root/casa_distro:/casa/setup ./brainvisa/casa-dev-5.0-11.sif branch=master distro=bvdev**
I have the follow error:

_FATAL:   container creation failed: mount /proc/self/fd/3->/var/lib/singularity/mnt/session/rootfs error: while mounting image /proc/self/fd/3: squashfs filesystem seems not enabled and/or supported by your kernel_

Any suggestion how I can solve this problem?

Best wishes,
Mike

**Comment by sapetnioc** [comment]
This error is on the side of Singularity. It has never been reported to us before. I do not know how to solve it. I saw on the web that it could be related to a bug in Singularity : [check this question](https://github.com/apptainer/singularity/issues/5408). Could you check your singularity version with `singularity --version` ?

**Reply by ece7048** [reply]
I had found this too. :) 

Mine is the suggested one form brainvisa for ubundu 18.03 that Colab runs:
singularity-ce_3.8.3_ubuntu-18.04_amd64.deb
so is v3.8.3

**Comment by sapetnioc** [comment]
I do not know Google Colab but it seems that you are not the only one to have trouble when trying to use singularity in that environment: https://stackoverflow.com/questions/63477944/running-singularity-container-from-colab-notebook

**Reply by ece7048** [reply]
Yeap I know that. I found all these recourses.! Thank you. :)

**Comment by ece7048** [comment]
I am trying with casa-distro to solve the issue by using source code. 
However I cannot understand how to build the enviroment with **caso_distro** command...
Thank you !!!

**Comment by denisri** [comment]
Hi,
I don't know much about google collab either, but it's likely already a kind of virtual machine or container, which is not designed to run another container inside it. Here it's possibly a filesystem mount type from the google collab virtual environment which is not supported by singularity (not really sure). The extra layers of google collab, jupyter and python inside it probably add some complexity.
Can't you try it on a standard computer ? To build the casa_distro environment you need a system able to run singularity (or virtualbox), and in the case of singularity, it also needs python installed because the startup scripts are written in python (no fancy module needed, a base standard python install is enough). Then doing what you were doing is OK (the `singularity run ...` command).

**Reply by ece7048** [reply]
Thank you for your response. 
I need GPUs so I cannot install it locally. 
Is there any way to use i**mport anatomist.direct.api as ana**
without installing the brainvisa from scratch??

**Reply by denisri** [reply]
Not an easy one. We do not have pip packages, because there are a lot of C++ compiled libraries in our software environment, with python bindings. Binary compatibility with all systems is too difficult to ensure (we don't have a development and porting team), so nowadays the only official way to install brainvisa and anatomist is to use virtualization. It is theoretically possible to compile all of it outside of a virtual environment, but there are a set of precise dependencies (with precise libs versions) that have been installed in the virtual environments, and that you may not have (and which may be a bit difficult to install) on a different system.
Now I don't know if the problem in google colab is specific to singularity. If docker works, it is possible to convert brainvisa images to docker.

**Reply by ece7048** [reply]
Thank you!! 
How I can convert braivisa image in docker?

Best wishes,
Mike

**Reply by denisri** [reply]
There is a tool in the master branch of casa-distro (not in the released brainvisa distributions): `casa_distro_admin convert_image source=brainvisa-5.0.4.sif container_type=docker`.
In 2 words it consists in duming the singularity image as a squashfs image file (`singularity sif dump <num> image.sif > data.squashfs`), unpacking the squashfs image (`unsquashfs -dest data data.squashfs`), then building a docker image from it.

**Reply by ece7048** [reply]
I did the: casa_distro, pip instal ., when I am trying to run casa_distro_admin  it says that there is no command. 
The setup.py install the casa_distro command but it appears that did not install the casa_distro-admin. Any ideas please?

**Reply by denisri** [reply]
No, as I said it was not on the released version. It's on github.

**Reply by ece7048** [reply]
Yes. I download the github casa-distro-master: https://github.com/brainvisa/casa-distro. I did the pre-commit install, and after a pip install of the setup.py.  It builds the casa_distro. When I use from terminal casa_distro command it recognise it. If I use casa_distro_admin did not. Is casa_distro_admin a different executable than casa_distro. I did not see any building process for this command in the setup.py file.  Thank you a lot for your time and effort.

**Reply by ece7048** [reply]
When I am running :
**!python3 ./casa-distro-master/bin/casa_distro_admin convert_image source="/content/gdrive/MyDrive/GNN/GCN/brainvisa-5.0.4.sif" container_type="docker" convert_from="singularity" verbose=1**

I have the follow error:

  _File "./casa-distro-master/bin/casa_distro_admin", line 25, in <module>
    main()
  File "/content/gdrive/MyDrive/GNN/GCN/casa-distro-master/python/casa_distro/command.py", line 360, in main
    result = command(*args, **kwargs)
  File "/content/gdrive/MyDrive/GNN/GCN/casa-distro-master/python/casa_distro/admin_commands.py", line 1464, in convert_image
    image_id, msg = module.convert_image(base, metadata, output,
  File "/content/gdrive/MyDrive/GNN/GCN/casa-distro-master/python/casa_distro/docker.py", line 460, in convert_image
    return convert_singularity_image_to_docker(source, metadata, output,
  File "/content/gdrive/MyDrive/GNN/GCN/casa-distro-master/python/casa_distro/docker.py", line 530, in convert_singularity_image_to_docker
    image_tag = metadata['name']
KeyError: 'name'_

Any ideas why I have no metadata information??

Thank you for your time and effort.

**Reply by c-langlet** [reply]
Hi, you probably don't have the corresponding json file: https://brainvisa.info/download/brainvisa-5.0.4.sif.json
